### PR TITLE
Updating to TypeScript 2.4.2

### PR DESCRIPTION
**Pull request by dzearing**
Getting the TypeScript 2.4.2 bump ready to go;

* Removed example-component because it was unnecessary
* Removed styling demo pages because they aren't used and created a circular dependency
* Upgraded a number of things including TypeScript, TSLint, webpack, and various build tools
* Fixed numerous ts and lint breaks

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -34,5 +34,5 @@
   "typescript.format.insertSpaceAfterOpeningAndBeforeClosingJsxExpressionBraces": true,
   "editor.formatOnSave": true,
   "typescript.tsdk": "./common/temp/node_modules/typescript/lib",
-  "tslint.autoFixOnSave": true
+  "tslint.autoFixOnSave": false
```
Comment (christiango) [inline .vscode/settings.json]: what'd you do that for?

File: apps/fabric-website/src/components/PageHeader/PageHeader.tsx [hunk 2]
```diff
@@ -9,6 +9,8 @@ const FULL_HEADER_HEIGHT = 236;
 const ATTACHED_HEADER_HEIGHT = 128;
 
 export interface IPageHeaderProps {
```
Comment (christiango) [inline apps/fabric-website/src/components/PageHeader/PageHeader.tsx]: You could also do extends here

File: common/config/rush/pinned-versions.json [hunk 3]
```diff
@@ -4,8 +4,8 @@
   "react-dom": "15.4.2",
   "enzyme": "2.7.0",
   "webpack": "2.2.1",
-  "tslint": "4.0.2",
-  "tslint-microsoft-contrib": "4.0.0",
+  "tslint": "5.5.0",
+  "tslint-microsoft-contrib": "5.0.1",
   "karma-sinon-chai": "1.3.1",
   "typescript": "2.2.2",
```
Comment (christiango) [inline common/config/rush/pinned-versions.json]: Update Typescript pinned-version

File: .vscode/settings.json (hunk 1 above)
Comment (dzearing) [inline .vscode/settings.json]: I keep seeing mistakes accidentally introduced by this!

For example; I needed to define an `any` to clean up an ambiguous and missing typing. Saving the file converted it into `{}` which wasted time multiple times when I didn't realize it happened.

So I'm not quite sure this one is good in particular. Sometimes it's better to just fail on the lint at build so you know what you're doing wrong.

Comment (dzearing) [inline .vscode/settings.json]: There are a set of lint rules that I'm very ok with auto fixing. Like redundant blank lines and such, which are just tedious to fix. But I think a lot of tslint categories require some thought.

File: packages/example-app-base/src/components/ComponentPage/ComponentPage.tsx [hunk 4]
```diff
@@ -92,22 +92,22 @@ export class ComponentPage extends React.Component<IComponentPageProps, {}> {
     if (bestPractices && dos && donts) {
       links.push(
         <div className='ComponentPage-navLink' key='nav-link'>
-          <Link href={ this._baseUrl + '#Best Practices' }>Best Practices</Link>
+          <Link { ...{ href: this._baseUrl + '#Best Practices' } }>Best Practices</Link>
```
Comment (cschleiden) [inline packages/example-app-base/src/components/ComponentPage/ComponentPage.tsx]: Why was this required?

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.Props.ts [hunk 5]
```diff
@@ -15,7 +15,7 @@ export interface ICheckbox {
 /**
  * Checkbox properties.
  */
-export interface ICheckboxProps extends React.HTMLAttributes<HTMLElement | HTMLInputElement> {
+export interface ICheckboxProps extends React.ButtonHTMLAttributes<HTMLElement | HTMLInputElement> {
```
Comment (cschleiden) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.Props.ts]: Is Button correct here?

File: packages/office-ui-fabric-react/src/components/TextField/TextField.tsx [hunk 6]
```diff
@@ -367,6 +367,10 @@ export class TextField extends BaseComponent<ITextFieldProps, ITextFieldState> i
       return;
     }
 
+    if (Array.isArray(value)) {
```
Comment (cschleiden) [inline packages/office-ui-fabric-react/src/components/TextField/TextField.tsx]: why should value be an array here?

File: .vscode/settings.json (hunk 1 above)
Comment (ThomasMichon) [inline .vscode/settings.json]: That's bad if it is auto-fixing semantic errors. It should only fix formatting and some syntax errors.

File: packages/office-ui-fabric-react/src/components/TextField/TextField.Props.ts [hunk 7]
```diff
@@ -95,13 +95,13 @@ export interface ITextFieldProps extends React.HTMLAttributes<HTMLInputElement |
    * Default value of the textfield, if any. Only provide this if the textfield is an uncontrolled component;
    * otherwise, use the "value" property.
    */
-  defaultValue?: string;
+  defaultValue?: string | undefined;
```
Comment (cschleiden) [inline packages/office-ui-fabric-react/src/components/TextField/TextField.Props.ts]: we need undefined now for an optional prop?

Comment (cschleiden) [inline packages/office-ui-fabric-react/src/components/TextField/TextField.Props.ts]: I think we don't need it, see: https://www.typescriptlang.org/play/index.html#src=interface%20X%20%7B%0D%0A%20%20%20%20foo%3F%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20Y%20%7B%0D%0A%20%20%20%20foo%3A%20string%3B%0D%0A%7D%0D%0A%0D%0A%0D%0Alet%20a%3A%20X%20%3D%20%7B%0D%0A%20%20%20%20foo%3A%20undefined%0D%0A%7D%3B%0D%0A%0D%0Alet%20b%3A%20Y%20%3D%20%7B%0D%0A%20%20%20%20foo%3A%20undefined%0D%0A%7D%3B%0D%0A

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.Props.ts (hunk 5 above)
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.Props.ts]: Technically yes, because we mix in the props to the button element. `HTMLAttribute` before meant "all html attributes", so technically we could go with that.

File: packages/example-app-base/src/components/ComponentPage/ComponentPage.tsx (hunk 4 above)
Comment (dzearing) [inline packages/example-app-base/src/components/ComponentPage/ComponentPage.tsx]: It's because example-app-base refers to the previous version of fabric (from npm) due to the circular dependency, and thus, the old link component is not compatible with the updated react typings. So as a temporary workaround until we're fully on the new typings, we have to work around it like this.
